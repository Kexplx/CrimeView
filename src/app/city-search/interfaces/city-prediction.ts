/**
 * Represents the prediction of a city-search, based on previous user input.
 *
 * "Reg" -> {placeId: "0129", name: "Regensburg"};
 */
export interface CityPrediction {
  placeId: string;
  name: string;
}
