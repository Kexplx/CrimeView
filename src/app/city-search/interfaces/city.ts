export interface Position {
  lat: number;
  lng: number;
}

export interface City {
  placeId: string;
  name: string;
  position: Position;
}
