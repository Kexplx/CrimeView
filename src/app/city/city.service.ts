import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

export interface CityPrediction {
  placeId: string;
  name: string;
}

export interface City {
  placeId: string;
  name: string;
  state: string;
  state_short: string;
  lat: number;
  lng: number;
}

@Injectable()
export class CityService {
  constructor(
    private predictionsService: google.maps.places.AutocompleteService,
    private geoCoder: google.maps.Geocoder,
  ) {}

  private predictionsOptions = {
    componentRestrictions: { country: 'de' },
    types: ['(cities)'],
  };

  getCityPredictions(input: string): Observable<CityPrediction[]> {
    return new Observable(sub => {
      this.predictionsService.getPlacePredictions(
        { ...this.predictionsOptions, input },
        predictions => {
          const cityPredictions = predictions.map<CityPrediction>(p => ({
            name: p.description,
            placeId: p.place_id,
          }));

          sub.next(cityPredictions);
        },
      );
    });
  }

  getCity(placeId: string): Observable<City> {
    return new Observable(sub => {
      this.geoCoder.geocode({ placeId }, result => {
        const { address_components, geometry } = result[0];

        const city: City = {
          placeId,
          lat: geometry.location.lat(),
          lng: geometry.location.lng(),
          name: address_components[0].long_name,
          state: address_components[3].long_name,
          state_short: address_components[3].short_name,
        };

        sub.next(city);
      });
    });
  }
}
