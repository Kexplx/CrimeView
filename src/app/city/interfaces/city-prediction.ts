/**
 * Represents the prediction of a city.
 *
 * "Reg" -> "Regensburg";
 */
export interface CityPrediction {
  placeId: string;
  name: string;
}
