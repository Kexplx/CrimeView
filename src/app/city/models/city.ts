export interface City {
  placeId: string;
  name: string;
  lat: number;
  lng: number;
}
