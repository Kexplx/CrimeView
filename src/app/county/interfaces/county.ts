import { Geometry } from 'geojson';

/**
 * The structure of a county.
 */
export interface County {
  name: string;
  state: string;
  type: string;
  countyCode: number;
  crimeRate: number | undefined;
  geometry: Geometry;
}
