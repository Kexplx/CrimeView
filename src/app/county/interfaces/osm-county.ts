import { Geometry } from 'geojson';

/**
 * The structure of a county oject returned by openstreetmap's api.
 */
export interface OsmCounty {
  name_1: string; // state, e.g. "Bayern"
  name_2: string; // name, e.g. "Regensburg"
  type_2: string; // type, e.g. "Landkreis", "Kreisfreihe Stadt"
  cca_2: string; // county code
  geo_shape: Geometry;
}
