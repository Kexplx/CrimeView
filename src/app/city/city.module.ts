import { NgModule } from '@angular/core';
import { CityListComponent } from './city-list/city-list.component';
import { SharedModule } from '../shared/shared.module';
import { CityService } from './city.service';
import { CityInputComponent } from './city-input/city-input.component';

@NgModule({
  declarations: [CityInputComponent, CityListComponent],
  imports: [SharedModule],
  providers: [
    CityService,
    google.maps.Geocoder,
    google.maps.places.AutocompleteService,
  ],
  exports: [CityListComponent],
})
export class CityModule {}
