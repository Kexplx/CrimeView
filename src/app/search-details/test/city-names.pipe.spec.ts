import { City } from '../../city-search/interfaces/city';
import { CityNamesPipe } from '../city-names.pipe';

const pipe = new CityNamesPipe();

it('should concat city-search names with "-" for type "Line"', () => {
  const cities: City[] = [{ name: 'Regensburg' } as City, { name: 'Muenchen' } as City];

  const result = pipe.transform(cities, 'Line');
  expect(result).toMatchSnapshot();
});

it('should concat city-search names with "-" for type "Polygon"', () => {
  const cities: City[] = [
    { name: 'Regensburg' } as City,
    { name: 'Muenchen' } as City,
    { name: 'Dortmund' } as City,
  ];

  const result = pipe.transform(cities, 'Polygon');
  expect(result).toMatchSnapshot();
});

it('should postfix first city-search name with "(10km Radius)" for type "Line"', () => {
  const cities: City[] = [{ name: 'Regensburg' } as City];

  const result = pipe.transform(cities, 'Radius');
  expect(result).toMatchSnapshot();
});
