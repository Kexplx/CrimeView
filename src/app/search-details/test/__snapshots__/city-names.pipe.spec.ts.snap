// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should concat city-search names with "-" for type "Line" 1`] = `"Regensburg - Muenchen"`;

exports[`should concat city-search names with "-" for type "Polygon" 1`] = `"Regensburg - Muenchen - Dortmund"`;

exports[`should postfix first city-search name with "(10km Radius)" for type "Line" 1`] = `"Regensburg (10km Radius)"`;
